File: src/services/ruleEngine.ts
/**
 * UserTrack Analytical Core & Rule Engine
 * Implements the algorithms from the specification:
 * 1. Brute Force Detector (5 errors in 5 min -> block 3 min, Critical incident)
 * 2. Time Anomaly Detector (2:00 <= hour <= 5:00 -> Medium incident)
 * 3. GeoIP Anomaly Detector (New IP / foreign country check -> Low/Medium incident)
 * 4. Honeypot Router Trap (/wp-login.php, /.env -> block 10 min, Critical incident, HACKER_BOT)
 * 5. Signature Analysis (Critical ports 22, 23, 3389, 3306, SQLi / attack patterns)
 * 6. IPS Active Prevention Layer (Drops blocked IPs before authentication)
 */

import { db } from './db';
import { NetworkEvent, SecurityIncident } from '../types';

export const HONEYPOT_PATHS = [
  { path: '/wp-login.php', desc: 'Фальшивая панель входа WordPress' },
  { path: '/.env', desc: 'Утечка файлов конфигурации и секретов' },
  { path: '/admin-panel-secret.php', desc: 'Скрытый административный бэкдор' },
  { path: '/phpmyadmin', desc: 'Интерфейс управления базами данных' },
  { path: '/api/v1/debug', desc: 'Неавторизованный отладочный endpoint' },
];

export const HONEYPOT_TRAPS = HONEYPOT_PATHS;

import { getQuickGeo, lookupIpGeo, GeoLookup } from './geoService';

export type { GeoLookup };

export function getIpInfo(ip: string): GeoLookup {
  return getQuickGeo(ip);
}

class RuleEngineService {
  /**
   * 1. Brute Force Protection (Сценарий 1)
   * Counts failures in the last 5 minutes. If >= 5, triggers IPS block for 3 mins.
   */
  public handleAuthAttempt(params: {
    ip: string;
    username: string;
    isSuccess: boolean;
    simulatedHour?: number; // allow testing night anomaly
    geoOverride?: GeoLookup;
  }): { allowed: boolean; blockedReason?: string; event: NetworkEvent; incident?: SecurityIncident } {
    const { ip, username, isSuccess, simulatedHour, geoOverride } = params;
    const geo = geoOverride || getIpInfo(ip);

    // If no geoOverride provided and geo is not yet finalized, resolve in background and patch event
    const patchEventGeoAsync = (eventId: number) => {
      if (!geoOverride && geo.country === 'Интернет') {
        lookupIpGeo(ip).then((realGeo) => {
          if (realGeo && realGeo.country && realGeo.country !== 'Интернет') {
            db.updateEvent(eventId, {
              country: realGeo.country,
              city: realGeo.city,
            });
          }
        }).catch(() => {});
      }
    };

    // Step 0: Check IPS block table first
    if (db.isIpBlocked(ip)) {
      const blockedEvent = db.addEvent({
        username,
        ipAddress: ip,
        eventType: 'SUSPICIOUS_PROBE',
        port: 443,
        country: geo.country,
        city: geo.city,
        details: `[IPS DROP] Запрос отклонён: IP ${ip} находится в черном списке блокировок`,
      });
      patchEventGeoAsync(blockedEvent.id);

      return {
        allowed: false,
        blockedReason: `IP-адрес ${ip} заблокирован системой IPS. Доступ запрещен до истечения таймаута.`,
        event: blockedEvent,
      };
    }

    if (!isSuccess) {
      // Record failure event
      const failEvent = db.addEvent({
        username,
        ipAddress: ip,
        eventType: 'AUTH_FAILURE',
        port: 443,
        country: geo.country,
        city: geo.city,
        details: `Неудачная попытка входа для пользователя '${username}'`,
      });
      patchEventGeoAsync(failEvent.id);

      // Count failures in the last 5 minutes
      const failCount = db.countRecentFailedLogins(ip, 5);

      if (failCount >= 5) {
        // Apply block on 5th failure
        db.applyBlock(ip, `Обнаружена Brute Force атака (${failCount} ошибок за 5 минут)`, 3, 'BRUTE_FORCE_BOT');
        const incident = db.addIncident({
          incidentType: 'BRUTE_FORCE',
          severity: 'Critical',
          description: `Зафиксирована серия из ${failCount} неудачных авторизаций с IP ${ip}. Система IPS заблокировала адрес на 3 минуты.`,
          sourceIp: ip,
          username,
          status: 'Active',
          ruleTriggered: 'RULE-BF-5MIN-THRESHOLD-5',
        });

        return {
          allowed: false,
          blockedReason: `Превышен лимит попыток входа (${failCount} ошибок). Ваш IP ${ip} автоматически заблокирован на 3 минуты модулем IPS!`,
          event: failEvent,
          incident,
        };
      }

      return {
        allowed: false,
        blockedReason: `Неверный пароль. Попытка ${failCount} из 5.`,
        event: failEvent,
      };
    }

    // AUTH SUCCESS
    const successEvent = db.addEvent({
      username,
      ipAddress: ip,
      eventType: 'AUTH_SUCCESS',
      port: 443,
      country: geo.country,
      city: geo.city,
      details: `Успешный вход пользователя '${username}' в систему`,
    });
    patchEventGeoAsync(successEvent.id);

    db.updateLastLogin(username);

    // 2. Check Time Anomaly (2:00 <= hour <= 5:00)
    const hour = simulatedHour !== undefined ? simulatedHour : new Date().getHours();
    let timeIncident: SecurityIncident | undefined;
    if (hour >= 2 && hour <= 5) {
      timeIncident = db.addIncident({
        incidentType: 'TIME_ANOMALY',
        severity: 'Medium',
        description: `Поведенческая аномалия: Вход пользователя '${username}' зафиксирован в аномальное ночное время (${hour}:15 MSK). Возможна работа бота или взлом учетной записи.`,
        sourceIp: ip,
        username,
        status: 'Active',
        ruleTriggered: 'UEBA-TIME-WINDOW-02-05',
      });
    }

    // 3. Check GeoIP Anomaly (New device / new IP)
    const user = db.getUserByUsername(username);
    let geoIncident: SecurityIncident | undefined;
    if (user && !user.knownIps.includes(ip)) {
      geoIncident = db.addIncident({
        incidentType: 'GEO_ANOMALY',
        severity: 'Low',
        description: `Вход с нового IP-адреса: Пользователь '${username}' ранее не авторизовывался с IP ${ip} (${geo.city}, ${geo.country}).`,
        sourceIp: ip,
        username,
        status: 'Active',
        ruleTriggered: 'UEBA-GEO-NEW-IP-DEVICE',
      });
      db.registerKnownIp(username, ip);
    }

    return {
      allowed: true,
      event: successEvent,
      incident: timeIncident || geoIncident,
    };
  }

  /**
   * 4. Honeypot Router Protection (Сценарий 2)
   * Instant 10 min block and Critical incident marked as HACKER_BOT.
   */
  public triggerHoneypot(path: string, ip: string, userAgent: string = 'Automated Scanner / Hydra'): {
    event: NetworkEvent;
    incident: SecurityIncident;
    block: unknown;
  } {
    const geo = getIpInfo(ip);

    const event = db.addEvent({
      username: 'HACKER_BOT',
      ipAddress: ip,
      eventType: 'HONEYPOT_HIT',
      port: 80,
      country: geo.country,
      city: geo.city,
      details: `[HONEYPOT TRAP] Попытка несанкционированного обращения к скрытой ловушке '${path}' (User-Agent: ${userAgent})`,
    });

    const block = db.applyBlock(
      ip,
      `Срабатывание ловушки Honeypot (запрос к фальшивому адресу '${path}')`,
      10,
      'HACKER_BOT'
    );

    const incident = db.addIncident({
      incidentType: 'HONEYPOT_TRIGGER',
      severity: 'Critical',
      description: `Сработал модуль контрразведки Honeypot: сканер уязвимостей попытался открыть '${path}' с IP ${ip} (${geo.country}). IP заблокирован на 10 минут, субъект классифицирован как HACKER_BOT.`,
      sourceIp: ip,
      username: 'HACKER_BOT',
      status: 'Active',
      ruleTriggered: `HONEYPOT-TRAP-${path.toUpperCase().replace(/[^A-Z0-9]/g, '_')}`,
    });

    return { event, incident, block };
  }

  /**
   * 5. Network Port & Signature Inspection (Сценарий 3 & Сканирование портов)
   */
  public inspectNetworkPacket(packet: {
    ip: string;
    port: number;
    protocol: 'TCP' | 'UDP' | 'HTTP' | 'HTTPS';
    payload: string;
    apiKey?: string;
  }): { event: NetworkEvent; incident?: SecurityIncident } {
    const { ip, port, protocol, payload, apiKey } = packet;
    const geo = getIpInfo(ip);

    // Signature checks
    const criticalPorts = [22, 23, 3389, 3306, 5432, 1433];
    const isCriticalPort = criticalPorts.includes(port);

    const isSqlInjection =
      /(\bUNION\b|\bSELECT\b|--|' OR '1'='1'|\bDROP\b|\bEXEC\b)/i.test(payload) ||
      payload.includes("1' OR '1' = '1");

    const isPathTraversal = payload.includes('../') || payload.includes('etc/passwd') || payload.includes('win.ini');

    let eventType: NetworkEvent['eventType'] = 'NETWORK_REQUEST';
    if (apiKey) {
      eventType = 'API_LOG_INGEST';
    } else if (isCriticalPort) {
      eventType = 'PORT_ACCESS';
    } else if (isSqlInjection || isPathTraversal) {
      eventType = 'SUSPICIOUS_PROBE';
    }

    const event = db.addEvent({
      username: apiKey ? 'external_agent.py' : 'anonymous',
      ipAddress: ip,
      eventType,
      port,
      country: geo.country,
      city: geo.city,
      protocol,
      details: apiKey
        ? `Внешний сбор телеметрии: ${payload}`
        : `Сетевой пакет на порт ${port}: ${payload}`,
      payloadSnippet: payload.substring(0, 80),
    });

    let incident: SecurityIncident | undefined;

    if (isCriticalPort) {
      incident = db.addIncident({
        incidentType: 'CRITICAL_PORT_ACCESS',
        severity: 'High',
        description: `Попытка несанкционированного подключения к критическому порту ${port} (${protocol}) с IP ${ip} (${geo.country})`,
        sourceIp: ip,
        status: 'Active',
        ruleTriggered: `SIG-RULE-PORT-${port}`,
      });
    } else if (isSqlInjection || isPathTraversal) {
      incident = db.addIncident({
        incidentType: 'MALICIOUS_PAYLOAD',
        severity: 'Critical',
        description: `Обнаружена сигнатура атаки (${isSqlInjection ? 'SQL Injection' : 'Path Traversal'}) в теле запроса с IP ${ip}`,
        sourceIp: ip,
        status: 'Active',
        ruleTriggered: isSqlInjection ? 'SIG-SQL-INJECTION' : 'SIG-PATH-TRAVERSAL',
      });
    }

    return { event, incident };
  }

  public ingestNetworkEvent(params: {
    ip: string;
    port: number;
    protocol?: 'TCP' | 'UDP' | 'HTTP' | 'HTTPS';
    details: string;
    country?: string;
    city?: string;
    username?: string;
  }): NetworkEvent {
    const res = this.inspectNetworkPacket({
      ip: params.ip,
      port: params.port,
      protocol: params.protocol || 'HTTP',
      payload: params.details,
    });
    return res.event;
  }
}

export const ruleEngine = new RuleEngineService();

File: backend/rule_engine.py
"""
UserTrack Analytical Core & UEBA / IPS Rule Engine (Python)
Ядро поведенческого анализа и модуль превентивных блокировок на чистом Python.
"""

from datetime import datetime
from typing import Dict, Any, Optional
from . import database as db
from . import geo_service

HONEYPOT_PATHS = [
    {"path": "/wp-login.php", "desc": "Фальшивая панель входа WordPress"},
    {"path": "/.env", "desc": "Утечка файлов конфигурации и секретов"},
    {"path": "/admin-panel-secret.php", "desc": "Скрытый административный бэкдор"},
    {"path": "/phpmyadmin", "desc": "Интерфейс управления базами данных"},
    {"path": "/api/v1/debug", "desc": "Неавторизованный отладочный endpoint"},
]

class RuleEngine:
    """Ядро правил безопасности UserTrack"""

    @staticmethod
    def handle_auth_attempt(
        ip: str,
        username: str,
        is_success: bool,
        simulated_hour: Optional[int] = None,
        geo_override: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Обработка попытки авторизации пользователя:
        1. Проверка активного бана в таблице blocks (IPS).
        2. При ошибке: фиксация AUTH_FAILURE, подсчет за 5 минут. При >= 5 ошибках — автобан на 3 мин.
        3. При успехе: фиксация AUTH_SUCCESS, проверка ночного входа (02:00-05:00), проверка нового IP (UEBA).
        """
        geo = geo_override or geo_service.get_quick_geo(ip)

        # 1. Проверка IPS черного списка
        if db.is_ip_blocked(ip):
            blocked_event = db.add_event({
                'username': username,
                'ipAddress': ip,
                'eventType': 'SUSPICIOUS_PROBE',
                'port': 443,
                'protocol': 'TCP',
                'country': geo.get('country', ''),
                'city': geo.get('city', ''),
                'details': f"[IPS DROP] Запрос отклонён: IP {ip} находится в активном черном списке блокировок"
            })

            return {
                'allowed': False,
                'blockedReason': f"IP-адрес {ip} заблокирован системой IPS. Доступ запрещен до истечения таймаута.",
                'event': blocked_event,
                'incident': None
            }

        # 2. Неудачная авторизация
        if not is_success:
            fail_event = db.add_event({
                'username': username,
                'ipAddress': ip,
                'eventType': 'AUTH_FAILURE',
                'port': 443,
                'protocol': 'TCP',
                'country': geo.get('country', ''),
                'city': geo.get('city', ''),
                'details': f"Неудачная попытка входа для пользователя '{username}' (неверный пароль)"
            })

            fail_count = db.count_recent_failed_logins(ip, minutes=5)

            if fail_count >= 5:
                # На 5-й ошибке активируем блокировку на 3 минуты
                db.apply_block(ip, f"Обнаружена Brute Force атака ({fail_count} ошибок за 5 минут)", duration_minutes=3, rule='BRUTE_FORCE_BOT')
                incident = db.add_incident({
                    'incidentType': 'BRUTE_FORCE',
                    'severity': 'Critical',
                    'description': f"Зафиксирована серия из {fail_count} неудачных авторизаций с IP {ip}. Модуль IPS заблокировал адрес на 3 минуты.",
                    'sourceIp': ip,
                    'username': username,
                    'status': 'Active',
                    'ruleTriggered': 'RULE-BF-5MIN-THRESHOLD-5'
                })

                return {
                    'allowed': False,
                    'blockedReason': f"Превышен лимит попыток входа ({fail_count} ошибок). Ваш IP {ip} автоматически заблокирован на 3 минуты модулем IPS!",
                    'event': fail_event,
                    'incident': incident
                }

            return {
                'allowed': False,
                'blockedReason': f"Неверный пароль. Попытка {fail_count} из 5.",
                'event': fail_event,
                'incident': None
            }

        # 3. Успешная авторизация (AUTH_SUCCESS)
        success_event = db.add_event({
            'username': username,
            'ipAddress': ip,
            'eventType': 'AUTH_SUCCESS',
            'port': 443,
            'protocol': 'TCP',
            'country': geo.get('country', ''),
            'city': geo.get('city', ''),
            'details': f"Успешный вход пользователя '{username}' в систему"
        })

        # Обновляем last_login пользователя в БД
        user = db.get_user_by_username(username)
        if user:
            db.update_user(user['id'], {'lastLogin': datetime.utcnow().isoformat() + 'Z'})

        incidents_created = []

        # Поведенческий анализ времени (UEBA: 02:00 - 05:00)
        hour = simulated_hour if simulated_hour is not None else datetime.now().hour
        if 2 <= hour <= 5:
            inc = db.add_incident({
                'incidentType': 'TIME_ANOMALY',
                'severity': 'Medium',
                'description': f"Поведенческая аномалия: Вход пользователя '{username}' зафиксирован в ночное нерабочее время ({hour}:15). Возможна работа бота или компрометация учетной записи.",
                'sourceIp': ip,
                'username': username,
                'status': 'Active',
                'ruleTriggered': 'RULE-TIME-NIGHT-LOGIN'
            })
            incidents_created.append(inc)

        # Поведенческий анализ геолокации и нового IP (UEBA Geo Anomaly)
        if user:
            known = user.get('knownIps', [])
            if ip not in known:
                country_name = geo.get('country', '')
                is_foreign = country_name and country_name not in ['Россия', 'Russian Federation', 'Локальная сеть (Loopback)', 'Корпоративная сеть (LAN)', '']
                severity = 'High' if is_foreign else 'Low'
                geo_desc = f"Вход с нового сетевого адреса {ip} ({geo.get('city', '')} {country_name})"
                if is_foreign:
                    geo_desc += f" — ВНИМАНИЕ: подключение из зарубежной локации ({country_name})!"

                inc = db.add_incident({
                    'incidentType': 'GEO_ANOMALY',
                    'severity': severity,
                    'description': f"Географическая аномалия: Пользователь '{username}'. {geo_desc}",
                    'sourceIp': ip,
                    'username': username,
                    'status': 'Active',
                    'ruleTriggered': 'RULE-GEO-UNRECOGNIZED-IP'
                })
                incidents_created.append(inc)

                # Добавляем IP в список доверенных
                updated_known = list(known) + [ip]
                db.update_user(user['id'], {'knownIps': updated_known})

        return {
            'allowed': True,
            'event': success_event,
            'incident': incidents_created[0] if incidents_created else None
        }

    @staticmethod
    def handle_honeypot_trap(path: str, ip: str, username: str = 'scanner_bot') -> Dict[str, Any]:
        """Срабатывание ловушки honeypot (краулер, сканирование секретов)"""
        geo = geo_service.get_quick_geo(ip)

        event = db.add_event({
            'username': username,
            'ipAddress': ip,
            'eventType': 'HONEYPOT_HIT',
            'port': 80,
            'protocol': 'HTTP',
            'country': geo.get('country', ''),
            'city': geo.get('city', ''),
            'details': f"[HONEYPOT TRIGGERED] Попытка несанкционированного доступа к скрытому маршруту: {path}"
        })

        # Мгновенная блокировка IPS на 10 минут
        block = db.apply_block(ip, f"Срабатывание Honeypot ловушки ({path})", duration_minutes=10, rule='HACKER_BOT')

        incident = db.add_incident({
            'incidentType': 'HONEYPOT_ACCESS',
            'severity': 'Critical',
            'description': f"Автоматическая реакция IPS: адрес {ip} обратился к защищенной ловушке {path}. Наложена немедленная блокировка на 10 минут.",
            'sourceIp': ip,
            'username': username,
            'status': 'Active',
            'ruleTriggered': 'RULE-HONEYPOT-ROUTER-TRAP'
        })

        return {
            'event': event,
            'block': block,
            'incident': incident
        }

    @staticmethod
    def handle_port_scan_probe(port: int, ip: str, username: str = 'network_probe') -> Dict[str, Any]:
        """Обработка сканирования портов и сигнатурного анализа"""
        geo = geo_service.get_quick_geo(ip)

        is_critical_port = port in [22, 23, 3389, 3306, 1433, 5432]
        port_names = {22: 'SSH', 23: 'Telnet', 3389: 'RDP', 3306: 'MySQL', 1433: 'MSSQL', 5432: 'PostgreSQL'}
        port_name = port_names.get(port, f"Port {port}")

        event = db.add_event({
            'username': username,
            'ipAddress': ip,
            'eventType': 'PORT_SCAN',
            'port': port,
            'protocol': 'TCP',
            'country': geo.get('country', ''),
            'city': geo.get('city', ''),
            'details': f"Сканирование сетевого порта {port} ({port_name})"
        })

        incident = None
        if is_critical_port:
            incident = db.add_incident({
                'incidentType': 'PORT_SCAN',
                'severity': 'High',
                'description': f"Обнаружено сканирование критического сервиса {port_name} (порт {port}) со стороны {ip}.",
                'sourceIp': ip,
                'username': username,
                'status': 'Active',
                'ruleTriggered': f"RULE-PORT-SCAN-{port}"
            })

        return {
            'event': event,
            'incident': incident
        }
